recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,190 @@
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
+  recipeConfig: {
+    terraform: {
+      authentication:{
+        ...        
+      }
++      providers: ['azurerm': {        
++          subscriptionid: 1234
++          secrets: {
++            source: secretStoreAz.id
++          }
++        },
++        'azurerm': {
++         subscriptionid: 1234
++         alias: 'az-paymentservice'
++         secrets: {
++           source: secretStoreAzPayment.id
++         }         
++       },
++       'gcp': {
++         project: 1234
++         regions: ['us-east1', 'us-west1']
++       },
++       'oraclepass': {
++         database_endpoint = "..."
++         java_endpoint     = "..."
++         mysql_endpoint    = "..."
++       }]
++     env: {
++       key: value
++       secrets: {
++        source: secretStoreOCICred.id
++      }
++     }
+    }
+  }
+    recipes: {      
+      ...
+    }
+  }
+}
+```
+
+
+Limitations: Customers may store sensitive data in other formats which may not be supported. eg. sensitive data is stored on files, which customers will not currently be able to load on disk in applications-rp where tf init/apply commands are run.
+Containerization work may alleviate this limitations. Further design work will be needed for towards this which is planned for the near future. 
+
+### API design (if applicable) (TBD)
+
+TBD
+
+## Alternatives considered
+
+Mentioned under Limitations above, work to containerize running terraform jobs was considered as a precursor to this work. The time sensitivity for unblocking customers on ability to configure and use providers they use today was given a priority and current design held. Containerization will be taken up as a parallel effort in the near future.
+
+## Test plan (TBD)
+
+#### Unit Tests (TBD)
+-   Update environment conversion unit tests to validate providers, env type under recipeConfig.
+-   Update environment controller unit tests for providers, env.
+
+#### Functional tests
+- Add e2e test to verify recipe using multiple provider configuration is deployed.
+- (TBD) Discuss/List providers we want to test with.
+
+## Security
+
+Largely following security considerations and secret handling described in design for private terraform modules: [Design document for Private Terraform Repository](https://github.com/radius-project/design-notes/blob/3644b754152edc97e641f537b20cf3d87a386c43/recipe/2024-01-support-private-terraform-repository.md)
+The work done here will be to read, decipher values and secrets and environment variables set by user in the *providers* block, to build internal terraform configuration file used by tf setenv, init and apply commands. 
+
+## Monitoring
+
+## Development plan (TBD)
+
+## Open issues/questions
+Do we consider first class support for GCP and other popular providers or continue with a generic approach??
```

Comment:
We should start with a generic approach to unlock everything, and then add first-class support where users request it later on. That we support everything, and then we can add convenience later on.